docs/advanced/life-cycle.md

```diff
@@ -70,3 +73,169 @@ export class FeatureModule{}
+  return () => Promise.resolve();
+}
+
+@NgModule({
+  providers: [
+    {
+      provide: APP_INITIALIZER,
+      useFactory: appInitializerFactory,
+      multi: true
+    }
+  ]
+})
+export class AppModule {}
+```
+
+This token is injected into `ApplicationInitStatus` class. What does Angular do under the hood? It gets an instance of this class and invokes the `runInitializers` method:
+
+```ts
+const initStatus = moduleRef.injector.get(ApplicationInitStatus);
+initStatus.runInitializers();
+```
+
+All that it does inside this method is looping via the `APP_INITIALIZER` array (as it's a `multi` token) and invoking those factories, that you've provided in `useFactory` properties:
+
+```ts
+for (let i = 0; i < this.appInits.length; i++) {
+  const initResult = this.appInits[i]();
+  if (isPromise(initResult)) {
+    asyncInitPromises.push(initResult);
+  }
+}
+```
+
+Then `asyncInitPromises` are provided into `Promise.all`. That's all the magic. That's why the `bootstrapModule` returns a `Promise`:
+
+```ts
+platformBrowser()
+  .bootstrapModule(AppModule)
+  .then(() => {
+    console.log('Hey there!');
+  });
+```
+
+### APP_INITIALIZER and NGXS
+
+Everything that we examined earlier is very important, because from this comes the fact that `APP_INITIALIZER` is resolved after NGXS states are initialized. They are initialized by the `NgxsModule` that is imported into the `AppModule`. The `ngxsOnInit` method on states is also invoked before the `APP_INITIALIZER` token is resolved. Given the following code:
+
+```ts
+@Injectable({ providedIn: 'root' })
+export class ConfigService {
+  private version: string | null = null;
+
+  constructor(private http: HttpClient) {}
+
+  loadVersion(): Observable<string> {
+    return this.http.get<string>('/api/version').pipe(
+      tap(version => {
+        this.version = version;
+      })
+    );
+  }
+
+  getVersion(): never | string {
+    if (this.version === null) {
+      throw new Error('"version" is not available yet!');
+    }
+
+    return this.version;
+  }
+}
+
+@State<string | null>({
+  name: 'version',
+  defaults: null
+})
+export class VersionState implements NgxsOnInit {
+  constructor(private configService: ConfigService) {}
+
+  ngxsOnInit(ctx: StateContext<string | null>) {
+    ctx.setState(this.configService.getVersion());
+  }
+}
+
+export function appInitializerFactory(configService: ConfigService) {
+  return () => configService.loadVersion().toPromise();
+}
+
+@NgModule({
+  imports: [NgxsModule.forRoot([VersionState])],
+  providers: [
+    {
+      provide: APP_INITIALIZER,
+      useFactory: appInitializerFactory,
+      multi: true,
+      deps: [ConfigService]
+    }
+  ]
+})
+export class AppModule {}
+```
+
+The above example is used only for demonstrating purposes! This code will throw an error because the `getVersion` method is invoked before the `version` property is set. Why? Because `ngxsOnInit` methods on states are invoked earlier!
```

Comment:
```suggestion
The above example is used only for demonstration purposes! This code will throw an error because the `getVersion` method is invoked before the `version` property is set. Why? Because the `ngxsOnInit` methods on states are invoked before the `APP_INITIALIZER` is invoked!
```